### ISSUE TITLE
Why not just using LSP4J?

### ISSUE THREAD

**Issue by mickaelistria** [scalameta/lsp4s#5]
I just stumbled upon this repository, and since there is already LSP4J which is working well and very well maintained and used by multiple clients and servers, I'm wondering what is the added value of LSP4S over LSP4J that justifies the effort ?

**Comment by gabro** [comment]
Hi @mickaelistria, we've initially tried lsp4j in [Metals](https://github.com/scalameta/metals/pull/31) (the project from which this library started), but failed to integrate it.

Essentially we've hit one roadblock after another when trying to use the API from Scala due to the heavy use of reflection in lsp4j which lead to cryptic errors.

Another minor reason, not compelling, was ergonomics: the API lsp4j is really not convenient to use from Scala (if we ever managed to).

Here's the relevant PR on Metals which hopefully provides some context: https://github.com/scalameta/metals/pull/31

**Comment by olafurpg** [comment]
haha, @gabro got ahead of me 😅 

Thanks for the question @mickaelistria ! We spent a couple days trying to get lsp4j working for https://github.com/scalameta/metals but the annotation-based runtime reflection was causing cryptic crashes. Writing the project in Scala ended up being easier for us to reason about and I think we also came up with a nice design

* no need to pass around cancellation token
* cancellation does not rely on exceptions, which can accidentally be caught
* easy to implement handlers for individual endpoints (like completions) without implementing the full `TextDocument` interface

Monix [`Observable`](https://monix.io/docs/3x/reactive/observable.html) is also a great fit for LSP in my experience.

If lsp4j works for you then by all means go ahead and use that project, I've heard it's good and well maintained.

**Comment by mickaelistria** [comment]
Ok, thanks for those answer.
cc @spoenemann

**Comment by olafurpg** [comment]
One more thing. An avenue we haven't explore yet with lsp4s but should be totally doable is to implement JavaScript clients with Scala.js. This might be handy when writing plugins for vscode/atom.

**Comment by laughedelic** [comment]
Probably you have something different in mind, but I think that lps4s on Scala.js wouldn't be of much help for the atom/vscode clients, because there are libraries like atom-languageclient and vscode-languageclient that implement the editor side of LSP and the language client plugin needs to use them, not the LSP interfaces directly. But anyway, some parts could be useful.

**Comment by olafurpg** [comment]
Good point @laughedelic, I haven't explored this avenue at all it was just an idea that popped up.

**Comment by olafurpg** [comment]
I investigated closer the option of using lsp4j instead of lsp4s and I have more details to share. In case anyone stumbles on this ticket.

First, I got it working! I encountered https://github.com/eclipse/lsp4j/issues/224 but worked around it by either
- extending `LanguageServer` in Java and writing some glue code so I can implement the endpoints in Scala 
- writing a custom language server protocol in Scala and using `Launcher` instead of `LSPLauncher`

However, if you have grown accustomed to writing libraries in Scala then the experience of writing a language server is still a bit nicer with lsp4s compared to lsp4j

- optional fields are represented as `null` in lsp4j, instead of `Option` in lsp4s
- endpoints with custom types require writing gson type adapters in lsp4j, instead of custom circe encoder/decoders in lsp4s
- lack of covariance means you end up with unnecessarily complex return types like `List<? extends Location>` in lsp4j instead of `List<Location>` in ls4ps
- endpoints return eager `CompletableFuture<T>` in lsp4j instead of lazy Monix `Task[T]` in lsp4s. Lazy tasks are pretty great in my experience since they enable non-invasive but fine-grained control over the execution context and cancellation comes out-of-the-box.

In a nutshell, if you want to write a language server in Scala with pure functional libraries and no runtime reflection then I recommend giving lsp4s a spin 😊 Otherwise, I think lsp4j is also a great choice!
